### ISSUE TITLE
Redirects and rewrites are basically the same thing

### ISSUE THREAD

**Issue by palant** [static-web-server/static-web-server#363]
### Search for duplicate feature request

- [X] I already searched, and this feature request or improvement is not a duplicate.

### Feature scope

Improve existing functionality

### Feature request related to a problem

Initially I didn’t recognize the difference between redirects and rewrites, and they are in fact almost identical. It seems that there are only two differences: redirects can be tied to a host name whereas as rewrites cannot, and rewrites support internal redirects whereas redirects don’t. It shouldn’t be too hard to merge both features and remove the confusion.

### Describe the solution you'd like

Adding an optional `host` field to rewrites will be backwards-compatible and render redirects superfluous. The redirects configuration options can stay for backwards compatibility but be rendered into `Rewrites` data structure. This will get rid of the confusion around these features and code duplication.

### Describe alternatives you've considered

Things can stay as they are now of course. Internal redirects will have to affect all hosts then.

### Build target

All targets

**Comment by joseluisq** [comment]
Although the idea is not bad. It would be great to see more details or a proposal for how It could look and would work using a single configuration for several use cases where both normally cover those scenarios separately.
I also wonder if such a merge might make the features harder to reason about for users.

Btw, if you asking yourself where the concept of redirects/rewrites in SWS are coming from then check out https://firebase.google.com/docs/hosting/full-config#rewrites

**Comment by palant** [comment]
> I also wonder if such a merge might make the features harder to reason about for users.

Well, we have that problem already. Rewrites can already be rewrites or redirects.

> Btw, if you asking yourself where the concept of redirects/rewrites in SWS are coming from then check out https://firebase.google.com/docs/hosting/full-config#rewrites

I see, I thought that it was something like that. And then you added redirect capabilities to rewrites and things got really muddy.

For reference, both nginx and Apache handle redirects and rewrites in the same directive – they merely have flags to turn an internal (default) into an external redirect. That’s essentially how SWS rewrites already behave as well.

**Comment by joseluisq** [comment]
> Well, we have that problem already. Rewrites can already be rewrites or redirects.

But not the other way around.
I wouldn't call it a problem but rather a different approach. The idea is to make it easier for users in general. So if the new approach result to be easier as well than the current one, then it could be the way to go.

> I see, I thought that it was something like that. And then you added redirect capabilities to rewrites and things got really muddy.

Since those actually are separated features, I do not see them as a problem honestly.

> For reference, both nginx and Apache handle redirects and rewrites in the same directive – they merely have flags to turn an internal (default) into an external redirect. That’s essentially how SWS rewrites already behave as well.

I have to look at a bit in detail.

Btw also take in to your account that the merge approach would be a breaking change. Because users would need to migrate/adjust their configurations.

**Comment by palant** [comment]
> Btw also take in to your account that the merge approach would be a breaking change. Because users would need to migrate/adjust their configurations.

No, my proposal specifically addresses this to keep existing configurations working.
